command/alloc_logs.go

```diff
@@ -275,30 +306,103 @@ func (l *AllocLogsCommand) followFile(client *api.Client, alloc *api.Allocation,
 
 	cancel := make(chan struct{})
 	frames, errCh := client.AllocFS().Logs(alloc, follow, task, logType, origin, offset, cancel, nil)
+
+	// Setting up the logs stream can fail, therefore we need to check the
+	// error channel before continuing further.
 	select {
 	case err := <-errCh:
 		return nil, err
 	default:
 	}
-	signalCh := make(chan os.Signal, 1)
-	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
 
-	// Create a reader
+	// Create a reader but don't initially cast it to an io.ReadCloser so that
+	// we can set the unblock time.
 	var r io.ReadCloser
 	frameReader := api.NewFrameReader(frames, errCh, cancel)
 	frameReader.SetUnblockTime(500 * time.Millisecond)
 	r = frameReader
 
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
+
+	// This go routine blocks until the command receives an interrupt or
+	// terminate signal, at which point we close the ReadCloser.
 	go func() {
 		<-signalCh
-
-		// End the streaming
-		r.Close()
+		_ = r.Close()
 	}()
 
 	return r, nil
 }
 
+// tailMultipleFiles will follow both stdout and stderr log files of the passed
+// allocation. Each stream will be output to the users console via stout and
+// stderr until the user cancels it.
+func (l *AllocLogsCommand) tailMultipleFiles(client *api.Client, alloc *api.Allocation) error {
+
+	// Use a single cancel channel for both log streams, so we only have to
+	// close one.
+	cancel := make(chan struct{})
+
+	// Ensure the channel is closed in order to notify listeners whenever we
+	// exit.
+	defer close(cancel)
+
+	stdoutFrames, stdoutErrCh := client.AllocFS().Logs(
+		alloc, true, l.task, api.FSLogNameStdout, api.OriginEnd, 1, cancel, nil)
```

Comment:
Reading the [docs](https://github.com/hashicorp/nomad/blob/43e25410d0bdd2ce0fc067bd4946cc50c9e54c58/api/fs.go#L238) I _think_ passing a `ctx.Done()` instead of creating a `cancel` channel would work, and it would make the code look more like modern Go.

You can then even pass the context to the function (`tailMultipleFiles(ctx, ...)`) and have let the parent `Run()` handle the lifecycle of the request, so that things like signal handling are handled outside of these sub-functions.

```go

func (l *AllocLogsCommand) Run(...) {
	// ...
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	// ...
	signalCh := make(chan os.Signal, 1)
	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
	go func() {
		for {
			select {
			case <-signalCh:
				return 0
			}
		}
	}()
	// ...
	tailMultipleFiles(ctx, ...)
	// ...
	followFile(ctx, ...)
}

func (l *AllocLogsCommand) tailMultipleFiles(ctx ...) error {
	// ...
	stderrFrames, stderrErrCh := client.AllocFS().Logs(
		alloc, true, l.task, api.FSLogNameStderr, api.OriginEnd, 1, ctx.Done(), nil)
	// ...
}

```